api-inference-community/docker_images/common/app/pipelines/structured_data_classification.py

```diff
@@ -0,0 +1,36 @@
+from typing import Dict, List, Union
+
+import joblib
+import numpy as np
+import pandas as pd
+from app.pipelines import Pipeline
+from huggingface_hub import cached_download, hf_hub_url
+
+
+class StructuredDataClassificationPipeline(Pipeline):
+    def __init__(self, model_id: str):
+        # IMPLEMENT_THIS
+        # Preload all the elements you are going to need at inference.
+        # For instance your model, processors, tokenizer that might be needed.
+        # This function is only called once, so do all the heavy processing I/O here
+        raise NotImplementedError(
+            "Please implement StructuredDataClassificationPipeline __init__ function"
+        )
+
+    def __call__(
+        self, inputs: Dict[str, Union[List[str], List[List[Union[str, float]]]]]
+    ) -> List[Union[str, float]]:
+        """
+        Args:
+            inputs (:obj:`dict`):
```

Comment:
This seems like a pretty ill defined pipeline API, in my eyes.

The end goal of `api-inference-community` is to propose a widget.
In order for this classification problem to work, the widget needs a way to understand expected prediction input vector X.

There is AFAIK, no way to have to get that on the widget side, they are stateless. So it would force the README to contain an example so the widget can display something usable (otherwise it's guesswork).

If we go that route, then we can limit the API here to a list of floats (maybe floats or strings for labels).  We need to emphasis that the implementor needs to do input sanitation too, and good error messages (limit the guesswork when examples are missing).

Optional arguments (like column names) do not really make sense IMO for a first pass on a Pipeline. 
Let's be overly explicit for now and require them (I guess that means input could become Dict[ColName, Value] for that, or ignore them entirely (they should be defined by the model anyway) and require List[Value].

Because of the aforementioned limits, I think much of the burden relies on the caller (the widget on this instant) to know how to organize the input, making this not very widget friendly.

Reply:
> In order for this classification problem to work, the widget needs a way to understand expected prediction input vector X.

This information could be in the metadata. #101 introduces a mixin that allows people to upload/load their pipelines. As discussed in #90 by @julien-c, we can populate some metadata (for example, `pipe.classes_`).

> There is AFAIK, no way to have to get that on the widget side, they are stateless. So it would force the README to contain an example so the widget can display something usable (otherwise it's guesswork).

My idea was to do a widget similar to `TableQuestionAnswering`. Users would fill a table and submit. The logic would be as follows:
* If column metadata in model card, we automatically fill the column names.
* If no column metadata, we show an empty grid. Users can or cannot specify column names (they are not really needed for predictions, but that might be too much guessing? In that case, maybe we can just provide the widget when there is column metadata available)

> Optional arguments (like column names) do not really make sense IMO for a first pass on a Pipeline. Let's be overly explicit for now and require them

I disagree here. Specially since when working with datasets with high-dimensionality users might have hundreds of columns and requiring column names might be annoying for users.